### PR TITLE
Update tag-details to handle foreign layers sanely (ie, to not invoke a HEAD request on them, which is likely to fail)

### PR DESCRIPTION
**Pull request by tianon**
cc @yosifkit

### CODE REVIEW

File: .tag-details.pl [hunk 1]
```diff
@@ -496,8 +500,9 @@ sub date {
 		say "-\t" . 'Layers:';
 		for my $layer (@{ $imageData->{layers} }) {
 			say "\t-\t" . '`' . $layer->{digest} . '`  ';
-			say "\t\t" . 'Last Modified: ' . date($layer->{lastModified}) . '  ';
-			say "\t\t" . 'Size: ' . size($layer->{size});
+			say "\t\t" . 'Last Modified: ' . date($layer->{lastModified}) . '  ' if defined $layer->{lastModified};
+			say "\t\t" . 'Size: ' . size($layer->{size}) . '  ' if defined $layer->{size};
+			say "\t\t" . 'MIME: ' . $layer->{mediaType};
```
Comment (tianon) [inline .tag-details.pl]: I figured we might as well include this now since it'll make it obvious when a layer is foreign. :smile:

**Comment by tianon** [comment]
Example output:

## `golang:windowsservercore`

```console
$ docker pull golang@sha256:36366df33f506c7f05d933cdb729e3432d026e02c7fa0d94a936cb945ef09230
```

-	Platforms:
	-	windows; amd64

### `golang:windowsservercore` - windows; amd64

-	Docker Version: 1.12.0
-	Manifest MIME: `application/vnd.docker.distribution.manifest.v2+json`
-	Total Size: **319.9 MB (319942528 bytes)**  
	(compressed transfer size, not on-disk size)
-	Image ID: `sha256:633190b5643a34196b3b2800c48593e3a32fda2d408af8130cb0d99dfc542574`
-	Default Command: `["c:\\windows\\system32\\cmd.exe"]`

```dockerfile
# Fri, 12 Aug 2016 17:01:50 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  SHELL [powershell -Command $ErrorActionPreference = 'Stop';]
# Fri, 12 Aug 2016 17:01:53 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_VERSION=2.9.2
# Fri, 12 Aug 2016 17:01:58 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_TAG=v2.9.2.windows.1
# Fri, 12 Aug 2016 17:02:02 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_DOWNLOAD_URL=https://github.com/git-for-windows/git/releases/download/v2.9.2.windows.1/Git-2.9.2-64-bit.exe
# Fri, 12 Aug 2016 17:03:48 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_DOWNLOAD_SHA256=006d971bcbe73cc8d841a100a4eb20d22e135142bf5b0f2120722fd420e166e5
# Fri, 12 Aug 2016 23:47:32 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; Write-Host ('Downloading {0} ...' -f $env:GIT_DOWNLOAD_URL); 	(New-Object System.Net.WebClient).DownloadFile($env:GIT_DOWNLOAD_URL, 'git.exe'); 		Write-Host ('Verifying sha256 ({0}) ...' -f $env:GIT_DOWNLOAD_SHA256); 	if ((Get-FileHash git.exe -Algorithm sha256).Hash -ne $env:GIT_DOWNLOAD_SHA256) { 		Write-Host 'FAILED!'; 		exit 1; 	}; 		Write-Host 'Installing ...'; 	Start-Process 		-Wait 		-FilePath ./git.exe 		-ArgumentList @( 			'/VERYSILENT', 			'/NORESTART', 			'/NOCANCEL', 		'/SP-', 			'/SUPPRESSMSGBOXES', 						'/COMPONENTS=assoc_sh', 						'/DIR=C:\git' 		); 		Write-Host 'Updating PATH ...'; 	$env:PATH = 'C:\git\bin;C:\git\mingw64\bin;C:\git\usr\bin;' + $env:PATH; 	[Environment]::SetEnvironmentVariable('PATH', $env:PATH, [EnvironmentVariableTarget]::Machine); 		Write-Host 'Verifying install ...'; 	Write-Host '  git --version'; git --version; 	Write-Host '  bash --version'; bash --version; 	Write-Host '  curl --version'; curl.exe --version; 		Write-Host 'Removing installer ...'; 	Remove-Item git.exe -Force; 		Write-Host 'Complete.';
# Fri, 12 Aug 2016 23:47:34 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOPATH=C:\gopath
# Fri, 12 Aug 2016 23:47:57 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; [Environment]::SetEnvironmentVariable('PATH', $env:GOPATH + '\bin;C:\go\bin;' + $env:PATH, [EnvironmentVariableTarget]::Machine);
# Tue, 16 Aug 2016 00:03:27 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOLANG_VERSION=1.7
# Tue, 16 Aug 2016 16:43:40 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOLANG_DOWNLOAD_URL=https://golang.org/dl/go1.7.windows-amd64.zip
# Tue, 16 Aug 2016 16:43:44 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOLANG_DOWNLOAD_SHA256=f51aad06644cc8bd119d2f6933334fa8da24d26e6676fde022cecf5978f1a0c7
# Tue, 16 Aug 2016 16:49:04 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; Write-Host ('Downloading {0} ...' -f $env:GOLANG_DOWNLOAD_URL); 	(New-Object System.Net.WebClient).DownloadFile($env:GOLANG_DOWNLOAD_URL, 'go.zip'); 		Write-Host ('Verifying sha256 ({0}) ...' -f $env:GOLANG_DOWNLOAD_SHA256); 	if ((Get-FileHash go.zip -Algorithm sha256).Hash -ne $env:GOLANG_DOWNLOAD_SHA256) { 		Write-Host 'FAILED!'; 		exit 1; 	}; 		Write-Host 'Expanding ...'; 	Expand-Archive go.zip -DestinationPath C:\; 		Write-Host 'Verifying install ("go version") ...'; 	go version; 		Write-Host 'Removing ...'; 	Remove-Item go.zip -Force; Write-Host 'Complete.';
# Tue, 16 Aug 2016 16:49:07 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  WORKDIR C:\gopath
```

-	Layers:
	-	`sha256:1239394e5a8ab79fbd3b751dc5d98decf5886f14339958fdf5c1f96c89da58a7`  
		Size: 3.5 GB (3461145128 bytes)  
		MIME: application/vnd.docker.image.rootfs.foreign.diff.tar.gzip
	-	`sha256:be0175dbdd12c150c89a273c6ae8a093858f186a1b130670b81917d39cfc62c3`  
		Last Modified: Tue, 16 Aug 2016 16:11:53 GMT  
		Size: 1.3 KB (1349 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:5a9f7fc8b1eb04be3a8b1a8fc1f9e6c7a1e41689a2676e152c25d9358dee1341`  
		Last Modified: Tue, 16 Aug 2016 16:11:51 GMT  
		Size: 1.4 KB (1362 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:94bdc4e8ef7ba6e9f01909b8cb91c6a122e96baef29babe69a01102bcae148c9`  
		Last Modified: Tue, 16 Aug 2016 16:11:47 GMT  
		Size: 1.3 KB (1347 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:45bbfdc9f3a15e9a0bde254b6ab26f6f076cf940ef39c7d7176f7642275bbfda`  
		Last Modified: Tue, 16 Aug 2016 16:11:47 GMT  
		Size: 1.3 KB (1345 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:8b64b13a7ee93c6f86303ad2fe1d9185da583cd6e7ab1671efe67b13e2993a11`  
		Last Modified: Tue, 16 Aug 2016 16:11:37 GMT  
		Size: 1.3 KB (1349 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:4e143173cb47f97febd951190e060d115fa15cf474acfb5759c8b410b45a1087`  
		Last Modified: Tue, 16 Aug 2016 16:15:14 GMT  
		Size: 222.7 MB (222712489 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:de436d4008dda1c26e300dd25d9cc67c8a59a2c0547febbcad41ae4499ea8882`  
		Last Modified: Tue, 16 Aug 2016 16:11:36 GMT  
		Size: 1.3 KB (1347 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:230b17b0e770bf3e266ac0f228d84bf32d942d15e3e5c57d4ec1e5061aa3e51e`  
		Last Modified: Tue, 16 Aug 2016 16:11:42 GMT  
		Size: 4.0 MB (4030539 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:2447de67ea2419aa2f8028e0fddf31710545f9ee8c9c813e1888f219a6614959`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.3 KB (1345 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:b2bbaa533ba7033fdaacd5a996517393d2c8220f7965dcbd6658800f6d1b75fe`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.4 KB (1351 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:02b47bb2bb68a606c91182bcd7724b7723f48341d6c24256e75c9b9d3a57b3f8`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.3 KB (1346 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:8835a8eeea92e980cb5e8a4deddb8dc46d36795c13a7aa2ff3a9c22a91e5c95d`  
		Last Modified: Tue, 16 Aug 2016 17:10:00 GMT  
		Size: 93.2 MB (93186010 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:a4af9de300cbd6f3f16e1daa9512a23f54638eb19f1db4a6f2da293febc25dea`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.3 KB (1349 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip

**Comment by tianon** [comment]
Slight fix in 758db5d results in: (including foreign layers in the total image size)

## `golang:windowsservercore`

```console
$ docker pull golang@sha256:36366df33f506c7f05d933cdb729e3432d026e02c7fa0d94a936cb945ef09230
```

-	Platforms:
	-	windows; amd64

### `golang:windowsservercore` - windows; amd64

-	Docker Version: 1.12.0
-	Manifest MIME: `application/vnd.docker.distribution.manifest.v2+json`
-	Total Size: **3.8 GB (3781087656 bytes)**  
	(compressed transfer size, not on-disk size)
-	Image ID: `sha256:633190b5643a34196b3b2800c48593e3a32fda2d408af8130cb0d99dfc542574`
-	Default Command: `["c:\\windows\\system32\\cmd.exe"]`

```dockerfile
# Fri, 12 Aug 2016 17:01:50 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  SHELL [powershell -Command $ErrorActionPreference = 'Stop';]
# Fri, 12 Aug 2016 17:01:53 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_VERSION=2.9.2
# Fri, 12 Aug 2016 17:01:58 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_TAG=v2.9.2.windows.1
# Fri, 12 Aug 2016 17:02:02 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_DOWNLOAD_URL=https://github.com/git-for-windows/git/releases/download/v2.9.2.windows.1/Git-2.9.2-64-bit.exe
# Fri, 12 Aug 2016 17:03:48 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GIT_DOWNLOAD_SHA256=006d971bcbe73cc8d841a100a4eb20d22e135142bf5b0f2120722fd420e166e5
# Fri, 12 Aug 2016 23:47:32 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; Write-Host ('Downloading {0} ...' -f $env:GIT_DOWNLOAD_URL); 	(New-Object System.Net.WebClient).DownloadFile($env:GIT_DOWNLOAD_URL, 'git.exe'); 		Write-Host ('Verifying sha256 ({0}) ...' -f $env:GIT_DOWNLOAD_SHA256); 	if ((Get-FileHash git.exe -Algorithm sha256).Hash -ne $env:GIT_DOWNLOAD_SHA256) { 		Write-Host 'FAILED!'; 		exit 1; 	}; 		Write-Host 'Installing ...'; 	Start-Process 		-Wait 		-FilePath ./git.exe 		-ArgumentList @( 			'/VERYSILENT', 			'/NORESTART', 			'/NOCANCEL', 		'/SP-', 			'/SUPPRESSMSGBOXES', 						'/COMPONENTS=assoc_sh', 						'/DIR=C:\git' 		); 		Write-Host 'Updating PATH ...'; 	$env:PATH = 'C:\git\bin;C:\git\mingw64\bin;C:\git\usr\bin;' + $env:PATH; 	[Environment]::SetEnvironmentVariable('PATH', $env:PATH, [EnvironmentVariableTarget]::Machine); 		Write-Host 'Verifying install ...'; 	Write-Host '  git --version'; git --version; 	Write-Host '  bash --version'; bash --version; 	Write-Host '  curl --version'; curl.exe --version; 		Write-Host 'Removing installer ...'; 	Remove-Item git.exe -Force; 		Write-Host 'Complete.';
# Fri, 12 Aug 2016 23:47:34 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOPATH=C:\gopath
# Fri, 12 Aug 2016 23:47:57 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; [Environment]::SetEnvironmentVariable('PATH', $env:GOPATH + '\bin;C:\go\bin;' + $env:PATH, [EnvironmentVariableTarget]::Machine);
# Tue, 16 Aug 2016 00:03:27 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOLANG_VERSION=1.7
# Tue, 16 Aug 2016 16:43:40 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOLANG_DOWNLOAD_URL=https://golang.org/dl/go1.7.windows-amd64.zip
# Tue, 16 Aug 2016 16:43:44 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  ENV GOLANG_DOWNLOAD_SHA256=f51aad06644cc8bd119d2f6933334fa8da24d26e6676fde022cecf5978f1a0c7
# Tue, 16 Aug 2016 16:49:04 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; Write-Host ('Downloading {0} ...' -f $env:GOLANG_DOWNLOAD_URL); 	(New-Object System.Net.WebClient).DownloadFile($env:GOLANG_DOWNLOAD_URL, 'go.zip'); 		Write-Host ('Verifying sha256 ({0}) ...' -f $env:GOLANG_DOWNLOAD_SHA256); 	if ((Get-FileHash go.zip -Algorithm sha256).Hash -ne $env:GOLANG_DOWNLOAD_SHA256) { 		Write-Host 'FAILED!'; 		exit 1; 	}; 		Write-Host 'Expanding ...'; 	Expand-Archive go.zip -DestinationPath C:\; 		Write-Host 'Verifying install ("go version") ...'; 	go version; 		Write-Host 'Removing ...'; 	Remove-Item go.zip -Force; Write-Host 'Complete.';
# Tue, 16 Aug 2016 16:49:07 GMT
RUN powershell -Command $ErrorActionPreference = 'Stop'; #(nop)  WORKDIR C:\gopath
```

-	Layers:
	-	`sha256:1239394e5a8ab79fbd3b751dc5d98decf5886f14339958fdf5c1f96c89da58a7`  
		Size: 3.5 GB (3461145128 bytes)  
		MIME: application/vnd.docker.image.rootfs.foreign.diff.tar.gzip
	-	`sha256:be0175dbdd12c150c89a273c6ae8a093858f186a1b130670b81917d39cfc62c3`  
		Last Modified: Tue, 16 Aug 2016 16:11:53 GMT  
		Size: 1.3 KB (1349 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:5a9f7fc8b1eb04be3a8b1a8fc1f9e6c7a1e41689a2676e152c25d9358dee1341`  
		Last Modified: Tue, 16 Aug 2016 16:11:51 GMT  
		Size: 1.4 KB (1362 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:94bdc4e8ef7ba6e9f01909b8cb91c6a122e96baef29babe69a01102bcae148c9`  
		Last Modified: Tue, 16 Aug 2016 16:11:47 GMT  
		Size: 1.3 KB (1347 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:45bbfdc9f3a15e9a0bde254b6ab26f6f076cf940ef39c7d7176f7642275bbfda`  
		Last Modified: Tue, 16 Aug 2016 16:11:47 GMT  
		Size: 1.3 KB (1345 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:8b64b13a7ee93c6f86303ad2fe1d9185da583cd6e7ab1671efe67b13e2993a11`  
		Last Modified: Tue, 16 Aug 2016 16:11:37 GMT  
		Size: 1.3 KB (1349 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:4e143173cb47f97febd951190e060d115fa15cf474acfb5759c8b410b45a1087`  
		Last Modified: Tue, 16 Aug 2016 16:15:14 GMT  
		Size: 222.7 MB (222712489 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:de436d4008dda1c26e300dd25d9cc67c8a59a2c0547febbcad41ae4499ea8882`  
		Last Modified: Tue, 16 Aug 2016 16:11:36 GMT  
		Size: 1.3 KB (1347 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:230b17b0e770bf3e266ac0f228d84bf32d942d15e3e5c57d4ec1e5061aa3e51e`  
		Last Modified: Tue, 16 Aug 2016 16:11:42 GMT  
		Size: 4.0 MB (4030539 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:2447de67ea2419aa2f8028e0fddf31710545f9ee8c9c813e1888f219a6614959`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.3 KB (1345 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:b2bbaa533ba7033fdaacd5a996517393d2c8220f7965dcbd6658800f6d1b75fe`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.4 KB (1351 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:02b47bb2bb68a606c91182bcd7724b7723f48341d6c24256e75c9b9d3a57b3f8`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.3 KB (1346 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:8835a8eeea92e980cb5e8a4deddb8dc46d36795c13a7aa2ff3a9c22a91e5c95d`  
		Last Modified: Tue, 16 Aug 2016 17:10:00 GMT  
		Size: 93.2 MB (93186010 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip
	-	`sha256:a4af9de300cbd6f3f16e1daa9512a23f54638eb19f1db4a6f2da293febc25dea`  
		Last Modified: Tue, 16 Aug 2016 17:09:11 GMT  
		Size: 1.3 KB (1349 bytes)  
		MIME: application/vnd.docker.image.rootfs.diff.tar.gzip

**Comment by yosifkit** [comment]
LGTM

**Pull request merged**